contrib/google.golang.org/grpc/appsec.go

```diff
@@ -9,24 +9,35 @@ import (
 	"encoding/json"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation/grpcsec"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation/httpsec"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo/instrumentation/sharedsec"
 
 	"github.com/DataDog/appsec-internal-go/netip"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
 )
 
 // UnaryHandler wrapper to use when AppSec is enabled to monitor its execution.
 func appsecUnaryHandlerMiddleware(span ddtrace.Span, handler grpc.UnaryHandler) grpc.UnaryHandler {
 	instrumentation.SetAppSecEnabledTags(span)
 	return func(ctx context.Context, req interface{}) (interface{}, error) {
+		var err error
 		md, _ := metadata.FromIncomingContext(ctx)
 		clientIP := setClientIP(ctx, span, md)
-		ctx, op := grpcsec.StartHandlerOperation(ctx, grpcsec.HandlerOperationArgs{Metadata: md, ClientIP: clientIP}, nil)
+		op := grpcsec.NewHandlerOperation(nil)
+		sharedsec.OnData(op, func(a *sharedsec.Action) {
+			code, e := a.GRPC()(md)
+			op.AddTag(instrumentation.BlockedRequestTag, true)
+			err = status.Error(codes.Code(code), e.Error())
+		})
+		ctx = grpcsec.StartHandlerOperation(ctx, op, grpcsec.HandlerOperationArgs{Metadata: md, ClientIP: clientIP})
```

Comment:
It took me a while to understand why you separated the operation creation from its start: to give a chance to subscribe to events before starting.

This is need indeed, but I suggest making this simpler with the `grpcsec.StartHandlerOperation` helper which is here to help making the integrations simpler to read and to avoid usage mistakes:

Something like:
```suggestion
		onAction := func(a *sharedsec.Action) {
			code, e := a.GRPC()(md)
			op.AddTag(instrumentation.BlockedRequestTag, true)
			err = status.Error(codes.Code(code), e.Error())
		}
		op, ctx = grpcsec.StartHandlerOperation(ctx, nil, grpcsec.HandlerOperationArgs{Metadata: md, ClientIP: clientIP}, onAction)
```